README.md
# Nuklear-Bindings : A Ruby bindings for Nuklear UI Library #

Provides Nuklear ( https://github.com/vurtun/nuklear ) interfaces for ruby (MRI).

*   Created : 2016-06-23
*   Last modified : 2017-08-14

## Prerequisites ##

*   Ruby-FFI ( https://github.com/ffi/ffi )
    *   $ gem install ffi

*   For demo/example.rb : opengl-bindings ( https://github.com/vaiorabbit/ruby-opengl )
    *   $ gem install opengl-bindings

## How to Use ##

1. Build Nuklear shared library
    *   See nuklear_dll

2. Include nuklear.rb in your script.
    *   ex.) require_relative 'nuklear'

3. Load shared library
    *   ex.) Nuklear.load_lib('libnuklear.dylib')

4. Setup OpenGL

See demo/example.rb for details.

## License ##

All source codes are available under the terms of the zlib/libpng license.

	Nuklear-Bindings : A Ruby bindings for Nuklear
	Copyright (c) 2016-2017 vaiorabbit
	
	This software is provided 'as-is', without any express or implied
	warranty. In no event will the authors be held liable for any damages
	arising from the use of this software.
	
	Permission is granted to anyone to use this software for any purpose,
	including commercial applications, and to alter it and redistribute it
	freely, subject to the following restrictions:
	
	    1. The origin of this software must not be misrepresented; you must not
	    claim that you wrote the original software. If you use this software
	    in a product, an acknowledgment in the product documentation would be
	    appreciated but is not required.
	
	    2. Altered source versions must be plainly marked as such, and must not be
	    misrepresented as being the original software.
	
	    3. This notice may not be removed or altered from any source
	    distribution.


ChangeLog
2017-08-14  vaiorabbit  <http://twitter.com/vaiorabbit>

	* Now compatible with nuklear bee90370281f294737f5b8ec6b5943dbf10b8e89 ( https://github.com/vurtun/nuklear/commit/bee90370281f294737f5b8ec6b5943dbf10b8e89 ).

2017-03-25  vaiorabbit  <http://twitter.com/vaiorabbit>

	* Now compatible with nuklear 944af0eb4cd02d6ec9f6527d3d698aed619ebe87 ( https://github.com/vurtun/nuklear/commit/944af0eb4cd02d6ec9f6527d3d698aed619ebe87 ).

2017-03-20  vaiorabbit  <http://twitter.com/vaiorabbit>

	* Now compatible with nuklear a979372909920a1800b0beaca3e967e7708b60bf ( https://github.com/vurtun/nuklear/commit/a979372909920a1800b0beaca3e967e7708b60bf ).

2016-11-28  vaiorabbit  <http://twitter.com/vaiorabbit>

	* demo: Added path settings for Ubuntu Linux.

2016-11-27  vaiorabbit  <http://twitter.com/vaiorabbit>

	* nuklear_dll/nuklear.def: Added module definition file for building Windows DLL.

2016-11-20  vaiorabbit  <http://twitter.com/vaiorabbit>

	* Now compatible with nuklear 6a374eabb1d8ace7e79c7562522cd435a3859393 ( https://github.com/vurtun/nuklear/commit/6a374eabb1d8ace7e79c7562522cd435a3859393 ).

2016-09-16  vaiorabbit  <http://twitter.com/vaiorabbit>

	* Now compatible with nuklear 3d3e63133e3deb58031a2c74a24d97eeff7b0d8d ( https://github.com/vurtun/nuklear/commit/3d3e63133e3deb58031a2c74a24d97eeff7b0d8d ).

2016-09-14  vaiorabbit  <http://twitter.com/vaiorabbit>

	* demo/overview.rb: Fixed some imcompatibility.

2016-09-11  vaiorabbit  <http://twitter.com/vaiorabbit>

	* Now compatible with nuklear@9e6b4f6 ( https://github.com/vurtun/nuklear/tree/9e6b4f67cc87145927fe6948f51d589cb0cfc424 ).

2016-09-10  vaiorabbit  <http://twitter.com/vaiorabbit>

	* demo/overview.rb: Now compatible with nuklear@8e4d8ee.
	* prev/nuklear.@8e4d8ee.h, prev/nuklear.@8e4d8ee.rb: Original code and wrapper for version 8e4d8ee ( https://github.com/vurtun/nuklear/tree/8e4d8ee741cf7559d98c9c43e1c64faa21fd593f ).

2016-08-27  vaiorabbit  <http://twitter.com/vaiorabbit>

	* nuklear.rb: Now compatible with v1.10.

2016-08-19  vaiorabbit  <http://twitter.com/vaiorabbit>

	* demo/overview.rb: Added (work in progress).

2016-08-18  vaiorabbit  <http://twitter.com/vaiorabbit>

	* nuklear.rb (nk_rect, etc.): Fixed wrong return values of nk_rect series APIs'

2016-08-17  vaiorabbit  <http://twitter.com/vaiorabbit>

	* nuklear.rb (nk_font_atlas_add, etc.): Fiexed to return NK_FONT directly.
	* nuklear.rb (nk_draw_foreach, etc.): Now code blocks get NK_COMMAND directly.

2016-08-16  vaiorabbit  <http://twitter.com/vaiorabbit>

	* demo/canvas_gl3.rb: Added.
	* demo/nkglfw_gl2.rb, demo/example_gl2.rb: Added.
	* demo/data/GenShinGothic-Normal.ttf, demo/unicode_gl3.rb: Added.

2016-08-15  vaiorabbit  <http://twitter.com/vaiorabbit>

	* nuklear.rb: Now compatible with v1.092.

2016-08-14  vaiorabbit  <http://twitter.com/vaiorabbit>

	* demo/nkglfw_gl3.rb, demo/example_gl3.rb: Added.

2016-08-13  vaiorabbit  <http://twitter.com/vaiorabbit>

	* nuklear.rb: Now compatible with v1.091.

2016-06-28  vaiorabbit  <http://twitter.com/vaiorabbit>

	* nuklear.rb (load_lib) : Use 'output_error' to see APIs you failed to import.

2016-06-27  vaiorabbit  <http://twitter.com/vaiorabbit>

	* demo/example.rb: Added mouse support.

2016-06-26  vaiorabbit  <http://twitter.com/vaiorabbit>

	* demo/example.rb: Added.
	* First release.

2016-06-23  vaiorabbit  <http://twitter.com/vaiorabbit>

	* Project started.


demo/example.rb
# coding: utf-8
require 'opengl'
require 'glfw'
require_relative '../nuklear'

OpenGL.load_lib()
GLFW.load_lib()
Nuklear.load_lib('libnuklear.dylib')

include OpenGL
include GLFW
include Nuklear

# Press ESC to exit.
key_callback = GLFW::create_callback(:GLFWkeyfun) do |window_handle, key, scancode, action, mods|
  if key == GLFW_KEY_ESCAPE && action == GLFW_PRESS
    glfwSetWindowShouldClose(window_handle, 1)
  end
end

if __FILE__ == $0
  glfwInit()

  window = glfwCreateWindow( 640, 480, "Simple example", nil, nil )
  glfwMakeContextCurrent( window )
  glfwSetKeyCallback( window, key_callback )

  nulldev = NK_DRAW_NULL_TEXTURE.new
  ctx = NK_CONTEXT.new

  nk_init_default(ctx, nil)
  cmds = NK_BUFFER.new
  nk_buffer_init_default(cmds)

  # Font definition begin
  atlas = NK_FONT_ATLAS.new
  nk_font_atlas_init_default(atlas)
  nk_font_atlas_begin(atlas)

  # Load fonts you like
  roboto_font = nil
  File.open("../nuklear/extra_font/GenShinGothic-Normal.ttf", "rb") do |ttf_file|
#  File.open("../nuklear/extra_font/Roboto-Bold.ttf", "rb") do |ttf_file|
    ttf_size = ttf_file.size()
    ttf = FFI::MemoryPointer.new(:uint8, ttf_size)
    content = ttf_file.read
    ttf.put_bytes(0, content)
    roboto_font_ptr = nk_font_atlas_add_from_memory(atlas, ttf, ttf_size, 22, nil)
    roboto_font = NK_FONT.new(roboto_font_ptr)
  end

  # Font definition end
  w = ' ' * 4
  h = ' ' * 4
  image = nk_font_atlas_bake(atlas, w, h, NK_FONT_ATLAS_FORMAT[:NK_FONT_ATLAS_RGBA32])
  # Upload atlas
  font_tex = ' ' * 4
  glGenTextures(1, font_tex)
  glBindTexture(GL_TEXTURE_2D, font_tex.unpack('L')[0])
  glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_LINEAR)
  glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_LINEAR)
  glTexImage2D(GL_TEXTURE_2D, 0, GL_RGBA, w.unpack('L')[0], h.unpack('L')[0], 0, GL_RGBA, GL_UNSIGNED_BYTE, image)
  nk_font_atlas_end(atlas, nk_handle_id(font_tex.unpack('L')[0]), nulldev)
  if atlas[:default_font].null? == false
    fnt = NK_FONT.new(atlas[:default_font])
    nk_style_set_font(ctx, fnt[:handle])
  else
    nk_style_set_font(ctx, roboto_font[:handle])
  end


  background = nk_rgb(64, 140, 216) # (28,48,62)

  compression_property = FFI::MemoryPointer.new(:int32, 1)
  compression_property.put_int32(0, 20)

  difficulty_option = FFI::MemoryPointer.new(:int32, 1)
  difficulty_option.put_int32(0, 0)

  while glfwWindowShouldClose( window ) == 0
    glfwPollEvents()

    fb_width_ptr = ' ' * 8
    fb_height_ptr = ' ' * 8
    win_width_ptr = ' ' * 8
    win_height_ptr = ' ' * 8
    glfwGetFramebufferSize(window, fb_width_ptr, fb_height_ptr)
    fb_width = fb_width_ptr.unpack('L')[0]
    fb_height = fb_height_ptr.unpack('L')[0]

    glfwGetWindowSize(window, win_width_ptr, win_height_ptr)
    win_width = win_width_ptr.unpack('L')[0]
    win_height = win_height_ptr.unpack('L')[0]

    # Update

    fb_scale_x = fb_width/win_width.to_f
    fb_scale_y = fb_height/win_height.to_f

    nk_input_begin(ctx)
    if ctx[:input][:mouse][:grab] != 0
      glfwSetInputMode(window, GLFW_CURSOR, GLFW_CURSOR_HIDDEN)
    elsif ctx[:input][:mouse][:ungrab] != 0
      glfwSetInputMode(window, GLFW_CURSOR, GLFW_CURSOR_NORMAL)
    end

    cursor_x_ptr = ' ' * 8
    cursor_y_ptr = ' ' * 8
    glfwGetCursorPos(window, cursor_x_ptr, cursor_y_ptr)
    cursor_x = cursor_x_ptr.unpack('D')[0]
    cursor_y = cursor_y_ptr.unpack('D')[0]
    nk_input_motion(ctx, cursor_x.to_i, cursor_y.to_i)
    if ctx[:input][:mouse][:grabbed] != 0
      glfwSetCursorPos(window, ctx[:input][:mouse][:prev][:x], ctx[:input][:mouse][:prev][:y])
      ctx[:input][:mouse][:pos][:x] = ctx[:input][:mouse][:prev][:x]
      ctx[:input][:mouse][:pos][:y] = ctx[:input][:mouse][:prev][:y]
    end
    nk_input_button(ctx, NK_BUTTONS[:NK_BUTTON_LEFT], cursor_x, cursor_y, glfwGetMouseButton(window, ((GLFW_MOUSE_BUTTON_LEFT) == GLFW_PRESS) ? 1 : 0))
    nk_input_button(ctx, NK_BUTTONS[:NK_BUTTON_MIDDLE], cursor_x, cursor_y, glfwGetMouseButton(window, ((GLFW_MOUSE_BUTTON_MIDDLE) == GLFW_PRESS) ? 1 : 0))
    nk_input_button(ctx, NK_BUTTONS[:NK_BUTTON_RIGHT], cursor_x, cursor_y, glfwGetMouseButton(window, ((GLFW_MOUSE_BUTTON_RIGHT) == GLFW_PRESS) ? 1 : 0))
    nk_input_end(ctx)


    # 3D
    ratio = fb_width.to_f / fb_height.to_f
    glViewport(0, 0, fb_width, fb_height)
    glClear(GL_COLOR_BUFFER_BIT)
    glClearColor( background[:r] / 255.0, background[:g] / 255.0, background[:b] / 255.0, 1.0 )
    glMatrixMode(GL_PROJECTION)
    glLoadIdentity()
    glOrtho(-ratio, ratio, -1.0, 1.0, 1.0, -1.0)
    glMatrixMode(GL_MODELVIEW)

    glLoadIdentity()
    glRotatef(glfwGetTime() * 50.0, 0.0, 0.0, 1.0)

    glBegin(GL_TRIANGLES)
    glColor3f(1.0, 0.0, 0.0)
    glVertex3f(-0.6, -0.4, 0.0)
    glColor3f(0.0, 1.0, 0.0)
    glVertex3f(0.6, -0.4, 0.0)
    glColor3f(0.0, 0.0, 1.0)
    glVertex3f(0.0, 0.6, 0.0)
    glEnd()

    # 2D
    rect = NK_RECT.new
    rect[:x] = 50.0
    rect[:y] = 50.0
    rect[:w] = 230.0
    rect[:h] = 250.0
    begin
      # Setup widgets
      layout = NK_PANEL.new
      r = nk_begin(ctx, layout, "Nuklear Ruby Bindings", rect,
                   NK_PANEL_FLAGS[:NK_WINDOW_BORDER]|
                   NK_PANEL_FLAGS[:NK_WINDOW_MOVABLE]|
                   NK_PANEL_FLAGS[:NK_WINDOW_SCALABLE]|
                   NK_PANEL_FLAGS[:NK_WINDOW_MINIMIZABLE]|
                   NK_PANEL_FLAGS[:NK_WINDOW_TITLE])
      if r != 0
        # Setup Widgets Here
        nk_layout_row_static(ctx, 30, 80, 1)
        nk_button_label(ctx, "button")
        nk_layout_row_dynamic(ctx, 30, 2)
        if nk_option_label(ctx, "eash", (difficulty_option.get_int32(0) == 0) ? 1 : 0) != 0
          difficulty_option.put_int32(0, 0)
        end
        if nk_option_label(ctx, "hard", (difficulty_option.get_int32(0) == 1) ? 1 : 0) != 0
          difficulty_option.put_int32(0, 1)
        end

        nk_layout_row_dynamic(ctx, 25, 1)
        nk_property_int(ctx, "Compression:", 0, compression_property, 100, 10, 1)
        combo = NK_PANEL.new
        nk_layout_row_dynamic(ctx, 20, 1)
        nk_label(ctx, "background:", NK_TEXT_ALIGNMENT[:NK_TEXT_LEFT])
        nk_layout_row_dynamic(ctx, 25, 1)
        res = nk_combo_begin_color(ctx, combo, background, 400)
        if res != 0
          nk_layout_row_dynamic(ctx, 120, 1)
          background = nk_color_picker(ctx, background, NK_COLOR_FORMAT[:NK_RGBA])
          nk_combo_end(ctx)
        end
      end
      nk_end(ctx)
    end

    begin
      # Render
      glPushAttrib(GL_ENABLE_BIT|GL_COLOR_BUFFER_BIT|GL_TRANSFORM_BIT)
      glDisable(GL_CULL_FACE)
      glDisable(GL_DEPTH_TEST)
      glEnable(GL_SCISSOR_TEST)
      glEnable(GL_BLEND)
      glEnable(GL_TEXTURE_2D)
      glBlendFunc(GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA)

      glViewport(0,0,fb_width,fb_height)
      glMatrixMode(GL_PROJECTION)
      glPushMatrix()
      glLoadIdentity()
      glOrtho(0.0, fb_width, fb_height, 0.0, -1.0, 1.0)
      glMatrixMode(GL_MODELVIEW)
      glPushMatrix()
      glLoadIdentity()

      glEnableClientState(GL_VERTEX_ARRAY)
      glEnableClientState(GL_TEXTURE_COORD_ARRAY)
      glEnableClientState(GL_COLOR_ARRAY)
      begin
        vs = NK_DRAW_VERTEX.size
        vp = NK_DRAW_VERTEX.offset_of(:position)
        vt = NK_DRAW_VERTEX.offset_of(:uv)
        vc = NK_DRAW_VERTEX.offset_of(:col)

        config = NK_CONVERT_CONFIG.new
        config[:global_alpha] = 1.0
        config[:shape_AA] = NK_ANTI_ALIASING[:NK_ANTI_ALIASING_ON]
        config[:line_AA] = NK_ANTI_ALIASING[:NK_ANTI_ALIASING_ON]
        config[:circle_segment_count] = 22
        config[:curve_segment_count] = 22
        config[:arc_segment_count] = 22
        config[:null] = nulldev

        vbuf = NK_BUFFER.new
        ebuf = NK_BUFFER.new
        nk_buffer_init_default(vbuf)
        nk_buffer_init_default(ebuf)
        nk_convert(ctx, cmds, vbuf, ebuf, config)
        vertices = nk_buffer_memory_const(vbuf)
        glVertexPointer(2, GL_FLOAT, vs, (vertices+vp))
        glTexCoordPointer(2, GL_FLOAT, vs, (vertices+vt))
        glColorPointer(4, GL_UNSIGNED_BYTE, vs, (vertices+vc))

        offset = nk_buffer_memory_const(ebuf)
        begin
          # draw widgets here
          nk_draw_foreach(ctx, cmds) do |cmd_ptr|
            cmd = NK_DRAW_COMMAND.new(cmd_ptr)
            next if cmd[:elem_count] == 0
            glBindTexture(GL_TEXTURE_2D, cmd[:texture][:id])
            glScissor(
                (cmd[:clip_rect][:x] * fb_scale_x).to_i,
                ((fb_height - (cmd[:clip_rect][:y] + cmd[:clip_rect][:h])).to_i * fb_scale_y).to_i,
                (cmd[:clip_rect][:w] * fb_scale_x).to_i,
                (cmd[:clip_rect][:h] * fb_scale_y)).to_i
            glDrawElements(GL_TRIANGLES, cmd[:elem_count], GL_UNSIGNED_SHORT, offset);
            offset += (FFI.type_size(:ushort) * cmd[:elem_count]) # NOTE : FFI.type_size(:ushort) == size of :nk_draw_index
          end
        end
        nk_clear(ctx)
        nk_buffer_free(vbuf)
        nk_buffer_free(ebuf)
      end
      glDisableClientState(GL_VERTEX_ARRAY)
      glDisableClientState(GL_TEXTURE_COORD_ARRAY)
      glDisableClientState(GL_COLOR_ARRAY)

      glDisable(GL_CULL_FACE)
      glDisable(GL_DEPTH_TEST)
      glDisable(GL_SCISSOR_TEST)
      glDisable(GL_BLEND)
      glDisable(GL_TEXTURE_2D)

      glBindTexture(GL_TEXTURE_2D, 0)
      glMatrixMode(GL_MODELVIEW)
      glPopMatrix()
      glMatrixMode(GL_PROJECTION)
      glPopMatrix()
      glPopAttrib()
    end

    glfwSwapBuffers( window )
  end

  nk_buffer_free(cmds)
  nk_free(ctx)

  glfwDestroyWindow( window )
  glfwTerminate()
end
